ProjFS.Mac/PrjFSKextLogDaemon/PrjFSKextLogDaemon.cpp

```diff
@@ -204,29 +256,169 @@ static bool TryFetchAndLogKextHealthData(io_connect_t connection)
+    healthDataWriter.Add("RefreshRoot", healthData.totalRefreshRootForVnode);
+    healthDataWriter.Add("InvalidateRoot", healthData.totalInvalidateVnodeRoot);
+    WriteJsonToMessageListener(HealthMessageEventName, healthDataWriter);
+}
+
+static void CreatePipeToMessageListener()
+{
+    lock_guard<mutex> lock(s_messageListenerMutex);
+    
+    if (INVALID_SOCKET_FD != s_messageListenerSocket)
+    {
+        // Already connected
+        return;
+    }
+
+    int error = 0;
+
+    s_messageListenerSocket = socket(PF_UNIX, SOCK_STREAM, 0);
+    if (s_messageListenerSocket < 0)
+    {
+        error = errno;
+        os_log(s_daemonLogger, "CreatePipeToMessageListener: Failed to create socket, error: %d", error);
+        s_messageListenerSocket = INVALID_SOCKET_FD;
+        return;
+    }
+    
+    struct sockaddr_un socket_address;
+    memset(&socket_address, 0, sizeof(struct sockaddr_un));
+    
+    socket_address.sun_family = AF_UNIX;
+    size_t pathLength = MessageListenerSocketPath.length();
+    if (pathLength + 1 >= sizeof(socket_address.sun_path))
+    {
+        os_log(s_daemonLogger, "CreatePipeToMessageListener: Failed to copy socket path, insufficient buffer. Buffer size: %lu", sizeof(socket_address.sun_path));
+        goto ClosePipeAndCleanup;
+    }
+    
+    strlcpy(socket_address.sun_path, MessageListenerSocketPath.c_str(), sizeof(socket_address.sun_path));
+    
+    if(0 == connect(s_messageListenerSocket, (struct sockaddr *) &socket_address, sizeof(struct sockaddr_un)))
+    {
+        os_log(s_daemonLogger, "CreatePipeToMessageListener: Connected to message listener");
+        return;
+    }
+    
+    error = errno;
+    os_log(s_daemonLogger, "CreatePipeToMessageListener: Failed to connect socket, error: %d", error);
+    
+ClosePipeAndCleanup:
+    ClosePipeToMessageListener_Locked();
+}
+
+static void ClosePipeToMessageListener_Locked()
+{
+    if (INVALID_SOCKET_FD != s_messageListenerSocket)
+    {
+        close(s_messageListenerSocket);
+        s_messageListenerSocket = INVALID_SOCKET_FD;
+    }
+}
+
+static void WriteJsonToMessageListener(const string& eventName, const JsonWriter& jsonMessage)
+{
+    lock_guard<mutex> lock(s_messageListenerMutex);
+    
+    if (INVALID_SOCKET_FD == s_messageListenerSocket)
+    {
+        return;
+    }
+
+    JsonWriter fullMessageWriter;
+    fullMessageWriter.Add("version", PrjFSKextVersion);
+    fullMessageWriter.Add("providerName", "Microsoft.Git.GVFS");
+    fullMessageWriter.Add("eventName", "kext." + eventName);
+    fullMessageWriter.Add("payload", jsonMessage);
+    
+    string fullMessage = fullMessageWriter.ToString() + "\n";
+    size_t bytesWritten;
+    do
+    {
+        bytesWritten = write(s_messageListenerSocket, fullMessage.c_str(), fullMessage.length());
+    } while (bytesWritten == -1 && errno == EINTR);
+    
+    int error = errno;
+    if (bytesWritten != fullMessage.length())
```

Comment:
The mac version of GitService runs in byte/stream mode and not in dgram/message mode, so we can do the usual write-loop logic for short writes, rather than aborting if you want.